sequencer/src/block2.rs

```diff
@@ -139,17 +151,53 @@ impl QueryableBlock for BlockPayload {
         index: &Self::TransactionIndex,
     ) -> Option<(Self::Transaction, Self::InclusionProof)> {
         let vid = boilerplate::test_vid_factory(); // TODO temporary VID construction
-        let tx_range = self.get_tx_range(*index)?;
+
+        let (tx_range, tx_table_range_proof) = self.get_tx_range_with_proof(*index, &vid)?;
+
         Some((
             // TODO don't copy the tx bytes into the return value
             // https://github.com/EspressoSystems/hotshot-query-service/issues/267
             Transaction::new(crate::VmId(0), self.payload.get(tx_range.clone())?.to_vec()),
-            vid.payload_proof(&self.payload, tx_range).ok()?,
+            TxInclusionProof {
+                // TODO(795) every proof will contain a copy of the tx table len proof.
+                // Should we cache it in `BlockPayload` or something?
+                tx_table_len_proof: vid
+                    .payload_proof(&self.payload, 0..TxTableEntry::byte_len())
+                    .ok()?,
+                tx_table_range_proof,
+                tx_payload_proof: if tx_range.is_empty() {
+                    None
+                } else {
+                    vid.payload_proof(&self.payload, tx_range).ok()
+                },
+            },
         ))
     }
 }
 
-mod tx_table {
+type TxIndex = <BlockPayload as QueryableBlock>::TransactionIndex;
+
+// TODO upstream type aliases: https://github.com/EspressoSystems/jellyfish/issues/423
+type RangeProof =
+    SmallRangeProof<<UnivariateKzgPCS<Bls12_381> as PolynomialCommitmentScheme>::Proof>;
+
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct TxInclusionProof {
```

Comment:
Either in this PR or soon after, we should implement a method something like

```rust
impl TxInclusionProof {
  fn validate(&self, payload_comm: VidCommitment, tx: &Transaction) -> bool;
}
```

This will just help be sure we haven't missed anything in the proof generation that we would need in the verifier. Although with this change I don't think we have (except of course the namespace stuff)

Reply:
Done in https://github.com/EspressoSystems/espresso-sequencer/pull/796/commits/90d8ddd4b103dff730c7ff59fd4a1f18e87f29de .

The new `verify()` method has an obnoxious arg list. I left it that way on purpose for now so you can see all the pieces needed to verify a proof. Where should all these pieces go? eg. Should they all become fields of `TxInclusionProof`?

Reply:
@jbearer Thinking about this a bit more: the problem is that we need both range and payload bytes for *three* subproofs in a `TxInclusionProof`. That's a lot of cruft.

I could just add all this info to `TxInclusionProof` so that the verifier doesn't need to hang on to this info and pass it in a big ugly statement to `verify` but that would conflict with the design principle we discussed in https://github.com/EspressoSystems/jellyfish/pull/389#discussion_r1383801714 . If it's okay to clean it up here then it should also be okay to clean it up in jellyfish, too.

If we're not careful we could create a cascade of copies of all this metadata at each level of the stack: a copy in the jellyfish proof plus a copy somewhere in the sequencer (maybe in the `TxInclusionProof`?) to pass to the jellyfish verification, etc.

I'm still not convinced that jellyfish verification should require the range. If we eliminate this requirement from jellyfish then it would eliminate the need to store 3 ranges in the sequencer. Moreover, I think we could include the payload bytes for the tx table length and entry inside `TxInclusionProof`, which would eliminate the need to store those things separately in the sequencer. I guess my point is that all this cruft is really just implementation details of the proof and should not be exposed to the user. wdyt?

Reply:
I think the {`table_len`,`entry_start`,`entry_end`} bytes could all be part of the proof. The reason is that the verifier isn't actually interested in those things, they just want to know generally that the transaction bytes are correct, and those things are implementation details of the proof.

I think we can _probably_ also move `tx_payload_range` to the proof, since the caller is not likely interested in reading a transaction at any particular offset (how would they know?), they just want to know they have the transaction at the right index. So only `tx_index` needs to be a verifier input.

Eventually, we may create an even higher-level wrapper for the API endpoint that puts `tx_index` in the proof as well, since probably the end user queried by hash, and they can check that just by recomputing the hash of `tx`. But at the level of abstraction we're currently at, I think it makes sense to expose `tx_index`.

Reply:
Ok sounds good.

I really don't like that the range is duplicated in both the jellyfish proof and in `TxInclusionProof` and I think I finally have a good solution. We made a mistake including this range in the jellyfish proof. As a general rule, each piece of data should be in the statement *or* the proof but never both. You have argued that the range should be in the (jellyfish) statement, so it should be removed from the (jellyfish) proof. Indeed, the only use of this data in jellyfish is to check that it equals the range given by the user in the `Statement`---in other words, it's completely useless.

This is a trivial upstream change that will save a few bytes of duplicated data inside proofs. Make sense?

Reply:
Yeah, sound sgood